src/aics-image-viewer/components/ViewerStateProvider/index.tsx

```diff
@@ -105,6 +95,49 @@ const applyChangeToViewerSettings = <K extends keyof ViewerState>(
   }
 };
 
+type ChannelStateAction<K extends keyof ChannelState> =
+  | {
+      // Update a single setting for one channel, or uniformly for multiple channels
+      index: number | number[];
+      key: K;
+      value: ChannelState[K];
+    }
+  | {
+      // Update a single setting for all channels from an ordered array of values
+      index?: undefined;
+      key: K;
+      value: ChannelState[K][];
+    }
+  | {
+      // Initialize channel settings
+      index?: undefined;
+      key?: undefined;
+      value: ChannelState[];
+    };
```

Comment:
This is sort of in the Nit category but see what you think:
What I'm used to seeing in redux land is that you'd actually name these actions like `InitializeChannelSettingsAction` for the last one, and then call `channelDispatch(new InitializeChannelSettingsAction({args or payload}));`.  I might suggest that the code is a bit more readable at the point of call when you don't have to overload one single type with so many variants.

Then the reducer could check the type of the action it receives (handwaving? my memory is bit vague on this)... 

You could still have some notion of a ChannelStateAction as a base type which has a index,key,and value.

Reply:
Definitely worth making the type definition clearer by splitting action types into named variants - I've just pushed a commit that does this, check it out - but I'm not sure about how much benefit we might gain from a readability standpoint with the other changes.

As I have it set up, we're missing two major parts of the redux pattern: actions with named types at runtime, and action creators. The main thing you're asking for above is action creators, though you've written it as a class constructor when I believe they're generally implemented as plain old functions. To work with an example, using action creators would turn this definition:

```typescript
const applyColorPresets = useCallback((value: ColorArray[]): void => channelDispatch({ key: "color", value }), []);
```
...into something like...
```typescript
const applyColorPresets = useCallback((value: ColorArray[]): void => channelDispatch(createChannelSettingArrayUpdateAction("color", value)));
```

Trouble is, this pattern starts introducing some of the extra ceremony that people love to grumble about in vanilla redux. Next to the type definition for `ChannelSettingArrayUpdateAction` (added in my last commit), we'd need a definition for `createChannelSettingArrayUpdateAction`:

```typescript
type ChannelSettingArrayUpdateAction<K extends keyof ChannelState> = {
  key: K;
  value: ChannelState[K][];
};
const createChannelSettingArrayUpdateAction<K extends keyof ChannelState> = (
  key: K,
  value: ChannelState[K][],
): ChannelSettingArrayUpdateAction<K> = ({ key, value });
```

That's a lot of redundant type info for a function that does nothing more than bundle its arguments into an object.

The other thing we're missing, which your mention of "the reducer could check the type of the action it receives" above gets at, is explicit action types at runtime. In redux, all actions are required to have a `type` field which names the type as a string literal. React lets you use any old type you want for actions (the react type definitions file actually explicitly mentions this difference with redux). I decided not to use a `type` field in the set of channel state actions, again because it felt unnecessary. The set of actions we want to take with channel settings isn't particularly complex: basically just "initialize" and a couple variants of "update a single setting." The reducer currently does type-check between different kinds of action here, it just does it based on missing properties. E.g. if the action is missing a `key` property, then it's not trying to update a single setting, it's trying to replace the entire state with a new initial value. One of my goals here was treating redux-flavored code not as a pattern the whole app would adopt, but as a necessary bit of complexity in the internal workings of just this one component, so it made some sense to me to take this less explicit approach. The new types layout tries to makes it clearer that the reducer is type-checking between multiple variants of `ChannelStateAction` - let me know if it works.

The way I see it, I'm happy with a nicer types definition, but we could decide to do a couple more things for readability:

1. Just bite the bullet and add a `type` field. This could make things clearer in lieu of action creators, since you'd at least be naming the kind of action you're taking at the `dispatch` call site and naming the type you're checking for in the reducer explicitly rather than in a comment.
2. Your use of a class constructor above actually suggests some syntax we could maybe use to unify action definition and action creator definition, e.g.:
```typescript
class ChannelSettingArrayUpdateActionClass<K extends keyof ChannelState> {
  index?: never;
  constructor(public key: K, public value: ChannelState[K][]) {}
}
```
Then we can use `new` syntax as you suggested above. It's not necessarily traditional redux, but again, react lets us do whatever we want with this pattern in a way redux doesn't.

Reply:
Thanks for this great reply.  I feel like maybe the class thing you mention at the end might help readability a tiny bit without introducing too much silly boilerplate.  Just having names for the three variants does help but I think it would help even more at the point where you first produce the data values, in the dispatch calls.  But at the same time I think I would be fine with the code in its original state too.